switchboard/base.py

```diff
@@ -6,77 +6,77 @@
+        # { key: '000-abc', 'name': 'Jim', 'phone': '1235677890' }
 
+        mydict = ModelDict(Model)
+        mydict['000-abc']
+        >>> Model({ 'key': '000-abc', 'name': 'Jim', 'phone': '1234567890' }) #doctest: +SKIP
 
-class CachedDict(threading.local):
-    def __init__(self, timeout=30):
-        """
-        Not guaranteed to be called with expected c'tor args of
-        all usages (due to usage of threading.local)
-        """
-        cls_name = type(self).__name__
+    If you want to use another key besides ``key``, you may specify that in the
+    constructor:
 
-        self._cache = None
-        self._last_updated = None
-        self.timeout = timeout
-        self.cache = settings.SWITCHBOARD_CACHE
-        self.cache_key = cls_name
-        self.last_updated_cache_key = '%s.last_updated' % (cls_name,)
+        mydict = ModelDict(Model, key='phone')
+        mydict['1234567890']
+        >>> Model({ 'key': '000-abc', 'name': 'Jim', 'phone': '1234567890' }) #doctest: +SKIP
+
+    The ModelDict needs to be thread local so that information is not shared
+    across threads, e.g., requests.
+    """
+    def __init__(self, model, key='key', auto_create=False, *args, **kwargs):
+        self._key = key
+        self._model = model
+        self._auto_create = auto_create
 
     def __getitem__(self, key):
-        self._populate()
-        try:
-            value = self._cache[key]
-        except KeyError:
-            value = self.get_default(key)
-            if value is NoValue:
-                raise
-        except (TypeError, AttributeError):
-            log.exception('Unable to access the local cache')
-            value = self.get_default(key)
-            if value is NoValue:
-                raise KeyError(key)
-        return value
+        if self._auto_create:
+            instance = self._model.get_or_create(key)[0]
+        else:
+            instance = self._model.get(key)
+        if instance is None:
+            raise KeyError(key)
+        return instance
+
+    def __setitem__(self, key, instance):
+        if not hasattr(instance, 'key'):
+            instance.key = key
+        instance.save()
+
+    def __delitem__(self, key):
+        self._model.remove(key)
 
     def __len__(self):  # pragma: nocover
-        if self._cache is None:
-            self._populate()
-        return len(self._cache)
+        return self._model.count()
 
     def __contains__(self, key):  # pragma: nocover
-        self._populate()
-        return key in self._cache
+        return self._model.contains(key)
 
     def __iter__(self):  # pragma: nocover
-        self._populate()
-        return iter(self._cache)
+        return self._model.all()
 
     def __repr__(self):  # pragma: nocover
         return "<%s>" % (self.__class__.__name__)
 
     def iteritems(self):  # pragma: nocover
-        self._populate()
-        return self._cache.iteritems()
+        return zip(self.iterkeys(), self.itervalues())
```

Comment:
This isn't very efficient, since `iterkeys()` and `itervalues()` will each end up calling `self._model.all()`.  Pratically though, I'm not sure if this method is really used anywhere that would matter.